Mob HP/Stat revamp

Frankie-hz opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?
Excels to referenced for data gathered:
Stat Tables: https://docs.google.com/spreadsheets/d/1YBoveP-weMdidrirY-vPDzHyxbEI2ryECINlfCnFkLI/edit#gid=799333298
Mob HP Retail Data & Formula Calcs: https://docs.google.com/spreadsheets/d/1ZjIRPBmIN1_EGraLnVx1GxmjwW96aHPujH1yyLufogg/edit?usp=sharing

HP Data was captured with capture suite + https://github.com/InoUno/HPMon

- [x] Revamp Mob HP.

Captured HP (In sheet)
Ogrefly:
![image](https://user-images.githubusercontent.com/105882754/235378190-3d65ac89-279a-4504-9e09-6c16bec43c9b.png)

New HP:
Ogrefly:
![image](https://user-images.githubusercontent.com/105882754/235378212-124b82f8-105e-4ecb-8240-34c01329c8d9.png)
![image](https://user-images.githubusercontent.com/105882754/235378223-006d3988-2979-4034-a6f5-e31e5450e520.png)

Tonberry Slasher:
![image](https://user-images.githubusercontent.com/105882754/235378257-a7991242-e82c-410d-b425-d3a804b1cbb0.png)
![image](https://user-images.githubusercontent.com/105882754/235378263-df9b6671-e6e2-4591-abe5-37ddeb64b86d.png)

Previous HP Scaling: TLDR: It is made up. Was based off of a random grown value. This was causing mobs all across the board to have extremely incorrect HP values. (Flies in Zitah have 800 less HP than they should.)

Stats:
- [x] Correct Base stat Values (STR, VIT, etc.) (Formulas were already good - just changed the subjob level)
- [x] Correct base damage for monsters (+5 for ranged, +2 for >1 and 0 to mimic a -1 level mob)
- [x] Correct DEF Values (These were only missing FLOOR)
- [x] Correct ATT Values (Assigned to correct Rank of A+)
- [x] Correct EVA Values (Assigned to C rank + 200 post-eva + agi/2

[stat] = floor[family Stat] + floor[main job Stat] + floor[sub job Stat]

Level 20 Fly WAR/WAR
STR: Rank A MJ + Rank A SJ + Family Rank E
`FLOOR(5 + ((20-1) * .5))` + `FLOOR(5 + ((20-1) * .5)) / 4` + `FLOOR( 3 + ((20-1) * .3))`
`14 + 8 + 3 = 25`

All the base stats are calculated the same way for their respective rank.
![image](https://user-images.githubusercontent.com/105882754/235570202-1ff6cc17-6b07-4c7d-904f-8f38d626c6f5.png)

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->

## Steps to test these changes
!getstats
<!-- Clear and detailed steps to test your changes here -->

Frankie-hz committed:
Mob HP Scaling Revamp

Frankie-hz committed:
Mob Stat Revamp

- Fixes Mobs Evasion, Attack, Acc, Ratt, Racc, HP

Xaver-DaRed commented on src/map/utils/mobutils.cpp:
```diff
@@ -513,10 +517,18 @@ namespace mobutils
+        PMob->addModifier(Mod::RACC, GetBase(PMob, PMob->accRank)); // Base Ranged Accuracy for all mobs is Rank A+ but pull from DB for specific cases
+
+        // Only mobs in dynamis can parry
+        if (PMob->isInDynamis())
```
I know this PR isn't finshed, but if I can give my opinion, I would feel more confortable if mobs had a mobMod for "canParry".
Or even a database column if that aint posible.

My logic behind this is, for instance, easier to customize for server operators. Also, I'm sure retail mobs can parry in places other than dynamis.
For instance, maybe some helper NPCs in BCNMs/assaults, which are mobs at the end of the day. Or the Generals in Besieged

Frankie-hz replied:
Which one would you prefer? Either or should work. I tend to stay away from adding more columns to sql but that's just a personal preference.

Xaver-DaRed replied:
I would go with a modifier, personaly.

TeoTwawki replied:
I'll second the mobmod for this case. that follows the pattern we recently set for shield blocking too.

Frankie-hz replied:
I added the mod in. I'll update the rest of the pool mods if this looks good. Let me know

Xaver-DaRed replied:
Looks good to me

TeoTwawki commented on src/map/entities/battleentity.cpp:
```diff
@@ -878,9 +878,7 @@ void CBattleEntity::SetSLevel(uint8 slvl)
-        // Technically, we shouldn't be assuming mobs even have a ratio they must adhere to.
-        // But there is no place in the DB to set subLV right now.
-        m_slvl = (slvl > (m_mlvl >> 1) ? (m_mlvl == 1 ? 1 : (m_mlvl >> 1)) : slvl);
+        m_slvl = m_mlvl; // All mobs have a 1:1 ratio of MainJob/Subjob
```
curious how this will affect Maat - right now he has no subjob.

Frankie-hz replied:
I did a Maat stat audit here after the utils change: https://github.com/AirSkyBoat/AirSkyBoat/pull/2303 I can try to find the notes we used for the FTPs/stat ranks. I should have it somewhere.

Frankie-hz committed:
Adds CAN_PARRY Mod

Adds CAN_PARRY for mobs that are allowed to parry.

Frankie-hz committed:
Merge branch 'base' into mob_hp_revamp

Frankie-hz commented:
Update:
Mob HP is finished.
Mob base stats and ATT values are finished.

I'll try to write up *something* for how we got HP values but mightt take me a while to remember exactly what we did (It's been a while.) But its pretty much spot on - maybe between 20-100 HP if you get higher up on levels. In any case its as accurate we are gonna get unless someone else can find a different formula that works as close as this does.

Frankie-hz committed:
Mob Stat Fixes

+ Added clamp to DEF so it does not overflow
+ Fixed mob Evasion calcs. Mobs do not have SKILL_EVASION - it goes by base Rank C formula + agi/2
+ Adjusted !getstats to show more stat values

Xaver-DaRed commented on src/map/grades.cpp:
```diff
@@ -136,6 +136,34 @@ std::array<std::array<float, 4>, 8> StatScale = { {
+
+/************************************************************************
+ *                                                                        *
+ *                                                                        *
```
We could use a little description of what RBI is/means

Or remove the giant comment box

Xaver-DaRed commented on src/map/grades.cpp:
```diff
@@ -193,4 +221,25 @@ namespace grade
+
+    /************************************************************************
+     *                                                                        *
+     *                                                                        *
```
Same here

Xaver-DaRed commented on src/map/utils/battleutils.cpp:
```diff
@@ -365,6 +365,12 @@ namespace battleutils
+        if (level > 99)
+        {
+            level = 99;
+        }
```
What effect can this have on current existing mobs that are over lvl 99?

Frankie-hz replied:
Every single instance of `uint16 GetMaxSkill` has `mlvl > 99 ? 99 : mlvl` in it. I am not sure why this wasn't just put in this function initially. Maybe I am missing something on this.

TeoTwawki replied:
There's been a lack of info on mob stats above 99, so our predecessors band-aid'd with a good ol "we'll do this for now and fix properly it later". 

Even 76-99 was just repeating lv 74->75's stat gains for a long time.

This little check probably fixed a few out of bound stats I've seen happen around level 108 or so...

Frankie-hz replied:
My knowledge is limited on mobs level 99+. My understanding was they are level 99 with enhanced stats, which needs loads of testing to even confirm that. Once that information is available, knowing SE its probably a static linear formula like the rest of the game that you could just add on (for at least base stats). As far as the indexing goes for skill_cap sql that would need  a small re-write or modification once that is figured out. Nonetheless this will at least fix the current system we have in place. I can add a TODO: with this line for research on 99+ mobs which isnt implemented atm anyway.

Frankie-hz replied:
> 

76 to 99 should have the correct stats - the table does scale to 99

TeoTwawki replied:
> My understanding was they are level 99 with enhanced stats, 

they for sure do that with players pets.

With mobs you fight they actually are reporting higher levels in the packet data, but my suspicion is mobs have 2 "kinds" of level - the actual level and then the level they _say_ they are. Kind of like players that are "119" are still actually level 99. I had suspected this before item level was a thing, because of detectors up in sky and the fact mobs report a level is 2 different packets. I figure when we got item levels they just extended a version of the same thing over to players.

> I can add a TODO: with this line for research on 99+ mobs which isn't implemented atm anyway.

sounds good to me

Xaver-DaRed commented on src/map/utils/mobutils.cpp:
```diff
@@ -231,68 +203,103 @@ namespace mobutils
         ZONE_TYPE zoneType = PMob->loc.zone->GetType();
 
+        uint8 grade;
+        uint8 gradesj;
```
My last nitpick: We can use verbose varibale names.

I mean, i can understand its the grade used for subjob, in this case, but for the sake of everyone else that will look/use this file after us, the more descriptive a variable name is, the easier it is to understand.

We could use mJobGrade and sJobGrade in this cases, as an example.
Just an example though. any other name that is descriptive is fine.

Xaver-DaRed commented:
Looking good man

Frankie-hz commented:
Everything is returning correctly. MEVA values will be correct when I PR some magic fixes + res_rank tables. It is just Rank C Eva + dlvl but that explanation is for another PR.

TeoTwawki commented on scripts/commands/getstats.lua:
```diff
@@ -32,19 +32,24 @@ function onTrigger(player)
+    player:PrintToPlayer(string.format("Total Accuracy: %i ", target:getStat(xi.mod.ACC)), xi.msg.channel.SYSTEM_3)
+    player:PrintToPlayer(string.format("RAccuracy Base: %i ", target:getMod(xi.mod.RACC)), xi.msg.channel.SYSTEM_3)
+    player:PrintToPlayer(string.format("Total RAccuracy: %i ", target:getStat(xi.mod.RACC)), xi.msg.channel.SYSTEM_3)
+    player:PrintToPlayer(string.format("EVA Base: %i ", target:getMod(xi.mod.EVA)), xi.msg.channel.SYSTEM_3)
```
just a note for future if you'd like to make any additions: I had originally planed to have a another parameter on the command and split up into multiple optional pages, as it gets kinda spammy. So further moderation to this are quit welcome. 👍

TeoTwawki reviewed (approved):
seems alight to me, more eyes welcome tho since its such a fundamental thing we've been after for so long
